util/resolver/resolver.go

```diff
@@ -3,37 +3,104 @@ package resolver
 import (
 	"context"
 	"crypto/tls"
+	"crypto/x509"
+	"io/ioutil"
 	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
 	"time"
 
 	"github.com/containerd/containerd/remotes"
 	"github.com/containerd/containerd/remotes/docker"
+	"github.com/moby/buildkit/cmd/buildkitd/config"
 	"github.com/moby/buildkit/session"
 	"github.com/moby/buildkit/session/auth"
 	"github.com/moby/buildkit/util/tracing"
+	"github.com/pkg/errors"
 )
 
-type RegistryConf struct {
-	Mirrors   []string
-	PlainHTTP *bool
-	Insecure  *bool
-}
+func fillInsecureOpts(host string, c config.RegistryConfig, h *docker.RegistryHost) error {
+	tc, err := loadTLSConfig(c)
+	if err != nil {
+		return err
+	}
 
-func fillInsecureOpts(host string, c RegistryConf, h *docker.RegistryHost) {
 	if c.PlainHTTP != nil && *c.PlainHTTP {
 		h.Scheme = "http"
 	} else if c.Insecure != nil && *c.Insecure {
-		h.Client = &http.Client{
-			Transport: tracing.NewTransport(&http.Transport{TLSClientConfig: &tls.Config{InsecureSkipVerify: true}}),
+		if tc == nil {
+			tc = &tls.Config{}
 		}
+		tc.InsecureSkipVerify = true
 	} else if c.PlainHTTP == nil {
 		if ok, _ := docker.MatchLocalhost(host); ok {
 			h.Scheme = "http"
 		}
 	}
+
+	if tc != nil && h.Scheme != "http" {
+		h.Client = &http.Client{
+			Transport: tracing.NewTransport(&http.Transport{TLSClientConfig: tc}),
+		}
+	}
+
+	return nil
+}
+
+func loadTLSConfig(c config.RegistryConfig) (*tls.Config, error) {
+	for _, d := range c.TLSConfigDir {
+		fs, err := ioutil.ReadDir(d)
+		if err != nil && !os.IsNotExist(err) && !os.IsPermission(err) {
+			return nil, errors.WithStack(err)
+		}
+		for _, f := range fs {
+			if strings.HasSuffix(f.Name(), ".crt") {
+				c.RootCAs = append(c.RootCAs, filepath.Join(d, f.Name()))
+			}
+			if strings.HasSuffix(f.Name(), ".cert") {
+				c.KeyPairs = append(c.KeyPairs, config.TLSKeyPair{
+					Certificate: filepath.Join(d, f.Name()),
+					Key:         filepath.Join(d, f.Name()[:len(f.Name())-5]+".key"),
+				})
```

Comment:
Needs documentation for this file name convention (both in code comments and in README.md)

Reply:
I copied Docker's behavior. Could even name it to `dockertlsconfig` maybe. Just to make sure we can use this when it is vendored to moby. @dmcgowan